experiment_analysis/isotonic.py

```diff
@@ -0,0 +1,70 @@
+from enum import Enum
+from typing import Union
+
+import numpy as np
+import pandas as pd
+from sklearn.isotonic import IsotonicRegression
+
+from experiment_analysis.exceptions import (
+    InvalidInputDataframe, InvalidIsotonicMethod
+)
+
+
+class IsotonicMethod(Enum):
+    STANDARD = "STANDARD"
+    DEMEAN = "DEMEAN"
+
+
+def apply_isotonic(
+    df: pd.DataFrame,
+    metric: str,
+    metric_cv: str,
+    method: Union[str, IsotonicMethod] = IsotonicMethod.STANDARD,
+) -> pd.Series:
+    """Apply Isotonic Regression to compute isotnoic metric based on experiment metric and pre-experiment covariate
+
+    :param df: pd.Dataframe
+    :param metric: str Name of dataframe column containing experiment metric
+    :param metric_cv: str Name of dataframe column containing pre-experiment covariates
+    :param method: Union[str, IsoraMethod] Variation on IsoraMethod method.
```

Comment:
Can you elaborate a little more on this here in the comment? What is the standard vs. demeaned version?

Reply:
So essentially isotonic regression and CUPED follow similar steps:
1. Fit a model to regress the **metric of interest** on **pre-xp covariates**, metric ~ pre-xp covariates
2. Calculate the residual
3. (Optional) Add back the mean of the metric

Similar to the implementation of `cuped.py`, the final step is actually optional since it doesn't change for group differences calculation since it's just a horizontal shift. [[source1](https://craft.faire.com/how-to-speed-up-your-a-b-test-outlier-capping-and-cuped-8c9df21c76b#:~:text=More%20specifically%2C%20we%20use%20a%20demeaned%20version%20of%20the%20covariate%20also%20used%20by%20Booking.com.%20Thus%2C%20the%20experiment%20metric%20after%20the%20CUPED%20transformation%20is%3A)] [[source2](https://booking.ai/how-booking-com-increases-the-power-of-online-experiments-with-cuped-995d186fff1d#:~:text=At%20Booking.com%2C%20we%20implement%20CUPED%20to%20adjust%20each%20unit%E2%80%99s%20metric%20score%20with%20the%20covariate%20as%20follows%20(which%20slightly%20differs%20from%20how%20Microsoft%20and%20Netflix%20use%20CUPED%20by%20including%20mean(covariate)%20to%20shift%20the%20sample%20mean%20for%20reporting%2C%20but%20this%20has%20no%20impact%20on%20group%20differences)%3A)]. But it does change lift calculation though due to the change in denominator. 

So all in all it's just aded as an option. Due to the fact that isotonic is more flexible than linear regression (run on CUPED) sometimes the demean variant might produce a tiny output metric that becomes unstable while calculating uplift, but for calculating pure differences it should be the same thing

Reply:
I think we can probably do without the demean version for the isotonic regression. The reason we have a demeaned version for CUPED is to match what's done in the XP pipelines. It's not better, it's just easier to then calculate percentage lifts. But the isotonic regression approach shouldn't have the issue that CUPED has where it substantially changes the mean. @yung-tang  let me know what you think, since you've been looking into this more recently

Reply:
@eramirezritchie Yes I think you're right about the fact that isotonic regression shouldn't substantially changes the mean. I do some quick checks on the sample XP we're looking at and for some cases using the demeaned version for isotonic regression actually gives us egregious output on lift calculation. I will remove the demean version shortly.